src/storage-client/src/types/sources/data.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! `SourceData` and related traits, e.g. other structs that can be taken to/from `SourceData`.
+
+use std::ops::{Deref, DerefMut};
+
+use anyhow::{anyhow, bail};
+use bytes::BufMut;
+use prost::Message;
+
+use mz_expr::PartitionId;
+use mz_persist_types::Codec;
+use mz_proto::{ProtoType, RustType, TryFromProtoError};
+use mz_repr::adt::range::RangeBoundDesc;
+use mz_repr::{Datum, Row};
+use mz_timely_util::order::{Interval, Partitioned, RangeBound};
+
+use crate::types::errors::DataflowError;
+
+use super::MzOffset;
+
+include!(concat!(
+    env!("OUT_DIR"),
+    "/mz_storage_client.types.sources.data.rs"
+));
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SourceData(pub Result<Row, DataflowError>);
+
+impl Deref for SourceData {
+    type Target = Result<Row, DataflowError>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for SourceData {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl RustType<ProtoSourceData> for SourceData {
+    fn into_proto(&self) -> ProtoSourceData {
+        use proto_source_data::Kind;
+        ProtoSourceData {
+            kind: Some(match &**self {
+                Ok(row) => Kind::Ok(row.into_proto()),
+                Err(err) => Kind::Err(err.into_proto()),
+            }),
+        }
+    }
+
+    fn from_proto(proto: ProtoSourceData) -> Result<Self, TryFromProtoError> {
+        use proto_source_data::Kind;
+        match proto.kind {
+            Some(kind) => match kind {
+                Kind::Ok(row) => Ok(SourceData(Ok(row.into_rust()?))),
+                Kind::Err(err) => Ok(SourceData(Err(err.into_rust()?))),
+            },
+            None => Result::Err(TryFromProtoError::missing_field("ProtoSourceData::kind")),
+        }
+    }
+}
+
+impl Codec for SourceData {
+    fn codec_name() -> String {
+        "protobuf[SourceData]".into()
+    }
+
+    fn encode<B: BufMut>(&self, buf: &mut B) {
+        self.into_proto()
+            .encode(buf)
+            .expect("no required fields means no initialization errors");
+    }
+
+    fn decode(buf: &[u8]) -> Result<Self, String> {
+        let proto = ProtoSourceData::decode(buf).map_err(|err| err.to_string())?;
+        proto.into_rust().map_err(|err| err.to_string())
+    }
+}
+
+impl From<Partitioned<PartitionId, MzOffset>> for SourceData {
```

Comment:
I suppose you moved this to a separate file due to this large `From` impl. Given that we'll delete this code very soon (after a couple of releases have been rolled out), should we keep the files as-is? We'll just need to suffer a slightly longer file for a few weeks

Reply:
Moved this back to the original location